Server/ConcordiaCurriculumManager/Services/CourseService.cs

```diff
@@ -89,32 +91,39 @@ public async Task<Course> GetCourseDataWithSupportingFilesOrThrowOnDeleted(strin
         var course = await _courseRepository.GetCourseWithSupportingFilesBySubjectAndCatalog(subject, catalog)
             ?? throw new InvalidInputException($"The course {subject}-{catalog} does not exist.");
 
-        if (course.CourseState == CourseStateEnum.Deleted)
-        {
-            throw new BadRequestException($"The course {subject}-{catalog} is deleted.");
+        if (course.CourseState == CourseStateEnum.Deleted)
+        {
+            throw new BadRequestException($"The course {subject}-{catalog} is deleted.");
         }
 
         return course;
-    }
-
-    public async Task<CourseCreationRequest> InitiateCourseCreation(CourseCreationInitiationDTO initiation, Guid userId)
-    {
-        var courseFromDb = await _courseRepository.GetCourseBySubjectAndCatalog(initiation.Subject, initiation.Catalog);
-        if (courseFromDb != null && courseFromDb.CourseState == CourseStateEnum.Accepted)
-        {
-            throw new BadRequestException("The course already exists and is accepted");
-        }
-
+    }
+
+    public async Task<CourseCreationRequest> InitiateCourseCreation(CourseCreationInitiationDTO initiation, Guid userId)
+    {
+        var courseFromDb = await _courseRepository.GetCourseBySubjectAndCatalog(initiation.Subject, initiation.Catalog);
+        if (courseFromDb != null && courseFromDb.CourseState == CourseStateEnum.Accepted)
+        {
+            throw new BadRequestException("The course already exists and is accepted");
+        }
+
+        var isDuplicate = await _dossierRepository.CheckIfCourseRequestExists(initiation.DossierId, initiation.Subject, initiation.Catalog);
+
+        if (isDuplicate)
+        {
+            throw new DuplicateNameException("A course creation request with the course " + initiation.Subject + " " + initiation.Catalog + " already exists.");

```

Comment:
The all 3 methods, (creation, modification, and deletion initiation), you should return a BadRequestException, because our backend is configured to return 400 level errors on those exceptions. Also, you should not specify `A course creation request with the course...` in the exception message, because you don't actually know what type of request it is (it might be a modification or deletion). This again applies to all 3 methods, just have the exception message say `"A course request for " + initiation.Subject + " " + initiation.Catalog + " already exists in this dossier."`

Reply:
Sounds good